### ISSUE TITLE
Feature Request SSH-Telnet-Serialport connection

### ISSUE THREAD

**Issue by nntantidakis** [microsoft/terminal#1280]
# Summary of the new feature/enhancement
It would be awesome if you could add tabs like ssh connection, Telnet connection and Serial port connection.

For example 
![image](https://user-images.githubusercontent.com/37312505/59552653-799f7000-8f92-11e9-83f9-a67ae6591321.png)
This is the list now on my PC, Imagine 3 new tabs with there 3 new options

Moreover, if you decide to implement the previous request, see if its easy to "remember" past connections, or better to save them with name, to show as tabs with the predefined connection details(of ssh and telnet)

**Comment by DHowett-MSFT** [comment]
This will best be handled by some combination of #555 and #626.

**Comment by DHowett-MSFT** [comment]
@hassanuz might be interested in this.

**Comment by DHowett-MSFT** [comment]
This is something you can do today with a bit of finagling; we'll be looking at what it means to remember SSH connections. The main point of contention is that we're not _a dedicated SSH client_ like putty and secureCRT are. Perhaps the Terminal isn't the right place for it?

**Comment by pilcherd** [comment]
I would like the ability to have a Serial Port terminal too.

I.e. open with baud rate/COM port settings etc.  

I know it feels so 1980's but in some fields there are many people who use it.  Putty supports a basic serial terminal so it's not that outside of the norm.

For all terminals, would like to be able to log terminal output to a file in different formats (raw/text) too.

**Comment by timoline** [comment]
IMHO, WT would be the perfect place for these kind of connections.
All in one solution

**Comment by nntantidakis** [comment]
pilcherd serial port connection is not 1980's. I mainly work with Wemos,Arduino etc and using serial port is extremely common. And the most frustating part of every program that reads serial connection is the fact that if the device disconnects, it closes the reader and they do not keep searching when the device will come online again for it to reopen automatically... So if you implement serial port terminal, please make it in a way that if your terminal detects a disconnection, not to close the tab-window but to inform the user that the device was disconnected,and reconnect automatically when the COM port becomes available again.

**Comment by pilcherd** [comment]
@nntantidakis I use serial ports in embedded device and instrumentation development daily.  My point was not that people don't use it today, but rather than most people will undervalue the importance of it for a select industries.  For most it feels an old or little used technology, but in the age of IoT it's growing in demand again.

**Comment by YaSuenag** [comment]
I hope that Windows Terminal provides features for serial console and SSH connection too. I think it is enough just a wrapper for them.

So I push them to my repository. They are very simple. I hope Windows Terminal will have similar features in the future.

* Serial console: https://github.com/YaSuenag/SimpleCom
* SSH client: https://github.com/YaSuenag/ssh-ps-wrapper

**Comment by rbharathkumar** [comment]
Considering the number of requests seen in the GitHub issues, I think the demand for Windows Terminal to come with its own ssh client is quite high, it would be really wonderful to include it. I believe the usage would skyrocket and bring in better feedback and request for features. I myself was looking to raise a similar issue but then saw the numerous others...
Thanks for all the hard work on the Windows Terminal.

**Comment by DHowett** [comment]
At least for now, *Windows* comes with a well-known and verifiable SSH client. Us relying on the one that comes with Windows means you don’t need to trust Terminal to be a security broker 😄

For now, that’s a better position for everyone to be in.

**Comment by samerhaj** [comment]
What about serial console manager? This issue seems to be tracking multiple requests. For me, a serial console manager would be an excellent addition to Windows Terminal, and removes the need to use additional programs just for serial connections to boards/devices.

**Comment by rbharathkumar** [comment]
> 
> 
> At least for now, _Windows_ comes with a well-known and verifiable SSH client. Us relying on the one that comes with Windows means you don’t need to trust Terminal to be a security broker 😄
> 
> For now, that’s a better position for everyone to be in.

Thank you for your response. I totally agree with you...One less thing to maintain and good part is that it is something we can rely and trust on...But there are some points to note though...

1. The OpenSSH Client that comes with Windows needs to be enabled from the Optional Windows Features meaning that if we do not have admin right which most corporate PCs and Laptops do not, we will not be able to enable it.
2. The OpenSSH Client option does not show up in some PCs like mine, I am currently on 1909 fully updated to date, I see the OpenSSH Server option but not the Client. I googled and found few other people having the same issue. (Not at all related to the Windows Terminal...but just FYI)
3. The installation instructions from the OpenSSH Portable github page seems a little complex with all its dependencies including a compiler and so on. Would be truly easier if Terminal came bundled with it or atleast enabled it via an option ?

I totally understand the logic behind the decision and Windows Terminal is not exactly geared towards that use case or the right tool for an SSH Client, but would be nice to have though..

Thank you once again.

**Comment by GitSheru** [comment]
> I would like the ability to have a Serial Port terminal too.
> 
> I.e. open with baud rate/COM port settings etc.
> 
> I know it feels so 1980's but in some fields there are many people who use it. Putty supports a basic serial terminal so it's not that outside of the norm.
> 
> For all terminals, would like to be able to log terminal output to a file in different formats (raw/text) too.

This should be a separate feature request. Much needed features

**Comment by GanterFengDavid** [comment]
3 years passed. Any progress on this feature request?

**Comment by zadjii-msft** [comment]
I don't suspect the core team will have the engineering resources to get to this any time soon. If I were someone really passionate about this, I'd probably:
* build another 3p app with UI for actually managing these connections. 
  * I'm not personally familiar enough with serial connections to know all the kinds of settings, bells and whistles that you may want to include in such a profile. But a dedicated app for that would be the fastest way to iterate on generating those profiles. 
* Use that app to generate [Fragment profiles](https://learn.microsoft.com/en-us/windows/terminal/json-fragment-extensions). 
  * The Terminal will automatically pick up profiles from fragment extensions. So your 3p app could generate profile blobs that can be used seamlessly with the Terminal. 

That could all be done today, without us needing to contribute to the core terminal codebase. Honestly, that's probably the best long-term solution to this.

**Comment by ManPython** [comment]
What a poor case..

**Comment by GanterFengDavid** [comment]
The [Tabby Terminal](https://github.com/Eugeny/tabby) can exactly meet this feature. It is also an alternative to Windows terminal.

**Comment by dlong500** [comment]
I agree with @zadjii-msft that what's really needed here is a third-party connection manager using Windows Terminal as the backend.  That would essentially allow for a PuTTY-like experience, but with all the benefit of the functionality and rapid iteration of Microsoft's standard Terminal app.

I understand that some people may need alternative terminals for a variety of technical reasons, but I think a lot of the compelling reasons for some third-party terminals would go away if we just had a good connection manager for the built-in Terminal app, and maintaining a connection manager would be orders of magnitude simpler than the development and maintenance of an entire third-party terminal app.

**Comment by bagobones** [comment]
Given that basically every network device in use has a serial/usb console connection needed for setup and debug, seems strange we can't just easily use the native terminal app.

Hell windows was bundled with Hyper Terminal that had this feature up till XP I think.
